pkg/storage/runtime.go

```diff
@@ -0,0 +1,415 @@
+		}
+		if err != nil {
+			return StorageContainerInfo{}, err
+		}
+	}
+	img, err := istorage.Transport.GetStoreImage(r.image.GetStore(), ref)
+	if img == nil && err == storage.ErrImageUnknown && metadata.ImageID != "" {
+		ref, err = istorage.Transport.ParseStoreReference(r.image.GetStore(), "@"+metadata.ImageID)
+		if err != nil {
+			return StorageContainerInfo{}, err
+		}
+		img, err = istorage.Transport.GetStoreImage(r.image.GetStore(), ref)
+	}
+	if err != nil && err != storage.ErrImageUnknown {
+		return StorageContainerInfo{}, err
+	}
+
+	// Pull the image down if we don't already have it.
+	if err == storage.ErrImageUnknown {
+		image := metadata.ImageID
+		if metadata.ImageName != "" {
+			image = metadata.ImageName
+		}
+		if image == "" {
+			return StorageContainerInfo{}, ErrInvalidImageName
+		}
+		logrus.Debugf("couldn't find image %q, retrieving it", image)
+		ref, err := r.image.PullImage(ctx, image)
+		if err != nil {
+			return StorageContainerInfo{}, err
+		}
+		img, err = istorage.Transport.GetStoreImage(r.image.GetStore(), ref)
+		if err != nil {
+			return StorageContainerInfo{}, err
+		}
+		logrus.Debugf("successfully pulled image %q", image)
+	}
+
+	// Fill in the image name and ID.
+	if metadata.ImageName == "" && len(img.Names) > 0 {
+		metadata.ImageName = img.Names[0]
+	}
+	metadata.ImageID = img.ID
+
+	// For now, we have a default cleanup function.
+	cleanup := func(container *storage.Container) {
+		if err2 := r.image.GetStore().DeleteContainer(container.ID); err2 != nil {
+			if metadata.Pod {
+				logrus.Infof("%v deleting partially-created pod sandbox %q", err2, container.ID)
+			} else {
+				logrus.Infof("%v deleting partially-created container %q", err2, container.ID)
+			}
+			return
+		}
+		logrus.Infof("deleted partially-created container %q", container.ID)
+	}
+
+	// Build metadata to store with the container.
+	mdata, err := json.Marshal(&metadata)
+	if err != nil {
+		return StorageContainerInfo{}, err
+	}
+
+	// Build the container.
+	names := []string{metadata.ContainerName}
+	if metadata.Pod {
+		names = append(names, metadata.PodName)
+	}
+	container, err := r.image.GetStore().CreateContainer(containerID, names, img.ID, "", string(mdata), nil)
+	if err != nil {
+		if metadata.Pod {
+			logrus.Debugf("failed to create pod sandbox %s(%s): %v", metadata.PodName, metadata.PodID, err)
+		} else {
+			logrus.Debugf("failed to create container %s(%s): %v", metadata.ContainerName, containerID, err)
+		}
+		return StorageContainerInfo{}, err
+	}
+	if metadata.Pod {
+		logrus.Debugf("created pod sandbox %q", container.ID)
+	} else {
+		logrus.Debugf("created container %q", container.ID)
+	}
+
+	// Add a name to the container's layer so that it's easier to follow
+	// what's going on if we're just looking at the storage-eye view of things.
+	layerName := metadata.ContainerName + "-layer"
+	names, err = r.image.GetStore().GetNames(container.LayerID)
+	if err != nil {
+		cleanup(container)
```

Comment:
Should the cleanup function be deferred instead so it is called on all subsequent error exits?